File: backend/app/agents/signal_classifier.py
"""
SignalClassifierAgent
Classifies news, filings, and other text into ESG or AI adoption signal categories.
"""
from typing import Dict, Any
import re


CATEGORY_RULES = {
    "environmental": [
        "emissions", "carbon", "climate", "pollution", "environmental", "renewable",
        "net zero", "biodiversity", "water", "waste", "spill", "deforestation",
    ],
    "social": [
        "employee", "worker", "safety", "diversity", "inclusion", "labor", "labour",
        "human rights", "community", "health", "discrimination", "harassment",
        "modern slavery", "supply chain",
    ],
    "governance": [
        "board", "executive", "corruption", "bribery", "audit", "shareholder",
        "whistleblower", "ethics", "compliance", "data privacy", "cybersecurity",
        "regulatory", "fine", "penalty", "lawsuit", "investigation",
    ],
    "ai_adoption": [
        "artificial intelligence", "machine learning", "automation", "AI strategy",
        "AI patent", "AI partnership", "digital transformation", "algorithm",
        "generative AI", "AI infrastructure", "robotics", "AI hiring",
    ],
    "controversy": [
        "scandal", "lawsuit", "fine", "violation", "greenwashing", "fraud",
        "accident", "disaster", "spill", "explosion", "death", "injury",
        "protest", "boycott", "recall",
    ],
}

SENTIMENT_RULES = {
    "positive": [
        "achieve", "reduce", "commit", "launch", "invest", "partner", "improve",
        "certif", "award", "recognize", "progress", "reach", "target met", "success",
        "breakthrough", "first", "lead",
    ],
    "negative": [
        "fail", "violat", "fine", "penalt", "lawsuit", "scandal", "miss", "decline",
        "increase emission", "accident", "death", "injury", "recall", "probe",
        "investigate", "allege", "charge",
    ],
}


def classify_text(title: str, body: str = "") -> Dict[str, Any]:
    text = (title + " " + body).lower()
    category_scores: Dict[str, int] = {k: 0 for k in CATEGORY_RULES}

    for cat, keywords in CATEGORY_RULES.items():
        for kw in keywords:
            if kw.lower() in text:
                category_scores[cat] += 1

    # Pick highest scoring category
    category = max(category_scores, key=lambda k: category_scores[k])
    if category_scores[category] == 0:
        category = "neutral"

    # Sentiment
    sentiment = "neutral"
    pos_score = sum(1 for w in SENTIMENT_RULES["positive"] if w in text)
    neg_score = sum(1 for w in SENTIMENT_RULES["negative"] if w in text)
    if pos_score > neg_score:
        sentiment = "positive"
    elif neg_score > pos_score:
        sentiment = "negative"

    # Severity for controversies
    severity = 0.0
    if category == "controversy" or sentiment == "negative":
        severity_keywords = ["death", "explosion", "disaster", "fraud", "billion", "criminal"]
        severity = min(10.0, 3.0 + sum(2.0 for w in severity_keywords if w in text))

    confidence = min(0.95, 0.5 + (category_scores.get(category, 0) * 0.08))

    return {
        "category": category,
        "sentiment": sentiment,
        "severity": severity,
        "confidence_score": round(confidence, 2),
    }


class SignalClassifierAgent:
    def classify(self, title: str, body: str = "", source: str = "", date: str = "") -> Dict[str, Any]:
        result = classify_text(title, body)
        explanation = self._generate_explanation(title, result)
        return {**result, "title": title, "source": source, "date": date, "explanation": explanation}

    def _generate_explanation(self, title: str, result: Dict[str, Any]) -> str:
        cat = result["category"]
        sent = result["sentiment"]
        sev = result["severity"]
        mapping = {
            "environmental": "This signal relates to environmental impact or climate action.",
            "social": "This signal relates to social factors including workforce and community.",
            "governance": "This signal relates to corporate governance, ethics or regulatory compliance.",
            "ai_adoption": "This signal indicates AI or digital transformation activity.",
            "controversy": "This signal flags a potential ESG controversy or risk event.",
            "neutral": "This signal has limited ESG relevance based on current analysis.",
        }
        base = mapping.get(cat, "Signal classified by keyword analysis.")
        if sent == "positive":
            base += " Sentiment is positive — potential improvement signal."
        elif sent == "negative":
            base += f" Sentiment is negative — severity rated {sev:.1f}/10."
        return base


signal_classifier_agent = SignalClassifierAgent()

File: backend-node/src/agents/signalClassifier.js
'use strict';

/**
 * SignalClassifierAgent — classifies ESG/AI signals from text using keyword rules.
 * Ported from Python with identical logic.
 */

const CATEGORY_RULES = {
  environmental: [
    'emissions', 'carbon', 'climate', 'pollution', 'environmental', 'renewable',
    'net zero', 'biodiversity', 'water', 'waste', 'spill', 'deforestation',
  ],
  social: [
    'employee', 'worker', 'safety', 'diversity', 'inclusion', 'labor', 'labour',
    'human rights', 'community', 'health', 'discrimination', 'harassment',
    'modern slavery', 'supply chain',
  ],
  governance: [
    'board', 'executive', 'corruption', 'bribery', 'audit', 'shareholder',
    'whistleblower', 'ethics', 'compliance', 'data privacy', 'cybersecurity',
    'regulatory', 'fine', 'penalty', 'lawsuit', 'investigation',
  ],
  ai_adoption: [
    'artificial intelligence', 'machine learning', 'automation', 'ai strategy',
    'ai patent', 'ai partnership', 'digital transformation', 'algorithm',
    'generative ai', 'ai infrastructure', 'robotics', 'ai hiring',
  ],
  controversy: [
    'scandal', 'lawsuit', 'fine', 'violation', 'greenwashing', 'fraud',
    'accident', 'disaster', 'spill', 'explosion', 'death', 'injury',
    'protest', 'boycott', 'recall',
  ],
};

const SENTIMENT_RULES = {
  positive: [
    'achieve', 'reduce', 'commit', 'launch', 'invest', 'partner', 'improve',
    'certif', 'award', 'recogni', 'progress', 'reach', 'target met', 'success',
    'breakthrough', 'first', 'lead',
  ],
  negative: [
    'fail', 'violat', 'fine', 'penalt', 'lawsuit', 'scandal', 'miss', 'decline',
    'increase emission', 'accident', 'death', 'injury', 'recall', 'probe',
    'investig', 'allege', 'charge',
  ],
};

function classifyText(title, body = '') {
  const text = `${title} ${body}`.toLowerCase();
  const categoryScores = Object.fromEntries(Object.keys(CATEGORY_RULES).map(k => [k, 0]));

  for (const [cat, keywords] of Object.entries(CATEGORY_RULES)) {
    for (const kw of keywords) {
      if (text.includes(kw.toLowerCase())) categoryScores[cat]++;
    }
  }

  let category = Object.entries(categoryScores).reduce((a, b) => a[1] >= b[1] ? a : b)[0];
  if (categoryScores[category] === 0) category = 'neutral';

  let sentiment = 'neutral';
  const posScore = SENTIMENT_RULES.positive.filter(w => text.includes(w)).length;
  const negScore = SENTIMENT_RULES.negative.filter(w => text.includes(w)).length;
  if (posScore > negScore) sentiment = 'positive';
  else if (negScore > posScore) sentiment = 'negative';

  let severity = 0.0;
  if (category === 'controversy' || sentiment === 'negative') {
    const severityKw = ['death', 'explosion', 'disaster', 'fraud', 'billion', 'criminal'];
    severity = Math.min(10.0, 3.0 + severityKw.filter(w => text.includes(w)).length * 2.0);
  }

  const confidence = Math.min(0.95, 0.5 + (categoryScores[category] || 0) * 0.08);

  return { category, sentiment, severity, confidenceScore: Math.round(confidence * 100) / 100 };
}

const CATEGORY_EXPLANATIONS = {
  environmental: 'This signal relates to environmental impact or climate action.',
  social: 'This signal relates to social factors including workforce and community.',
  governance: 'This signal relates to corporate governance, ethics or regulatory compliance.',
  ai_adoption: 'This signal indicates AI or digital transformation activity.',
  controversy: 'This signal flags a potential ESG controversy or risk event.',
  neutral: 'This signal has limited ESG relevance based on current analysis.',
};

function classify(title, body = '', source = '', date = '') {
  const result = classifyText(title, body);
  let explanation = CATEGORY_EXPLANATIONS[result.category] || 'Signal classified by keyword analysis.';
  if (result.sentiment === 'positive') explanation += ' Sentiment is positive — potential improvement signal.';
  else if (result.sentiment === 'negative') explanation += ` Sentiment is negative — severity rated ${result.severity.toFixed(1)}/10.`;

  return { ...result, title, source, date, explanation };
}

module.exports = { classify, classifyText };
